x-pack/plugins/security/public/components/form_flyout.tsx

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect, FunctionComponent, RefObject } from 'react';
+import { FormattedMessage } from '@kbn/i18n/react';
+import {
+  EuiTitle,
+  EuiFlyout,
+  EuiFlyoutProps,
+  EuiFlyoutHeader,
+  EuiFlyoutBody,
+  EuiFlyoutFooter,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiButton,
+  EuiButtonProps,
+  EuiButtonEmpty,
+  EuiPortal,
+} from '@elastic/eui';
+
+export interface FormFlyoutProps extends Omit<EuiFlyoutProps, 'onSubmit' | 'onClose'> {
+  title: string;
+  initialFocus?: RefObject<HTMLElement>;
+  onCancel(): void;
+  onSubmit(): void;
+  submitButtonText: string;
+  submitButtonColor?: EuiButtonProps['color'];
+  isLoading?: EuiButtonProps['isLoading'];
+  isDisabled?: EuiButtonProps['isDisabled'];
+}
+
+export const FormFlyout: FunctionComponent<FormFlyoutProps> = ({
+  title,
+  submitButtonText,
+  submitButtonColor,
+  onCancel,
+  onSubmit,
+  isLoading,
+  isDisabled,
+  children,
+  initialFocus,
+  ownFocus = true,
+  ...rest
+}) => {
+  useEffect(() => {
+    if (initialFocus && initialFocus.current) {
+      initialFocus.current.focus();
+    }
+  }, [initialFocus]);
+
+  const flyout = (
+    <EuiFlyout ownFocus={ownFocus} onClose={onCancel} {...rest}>
+      <EuiFlyoutHeader hasBorder>
+        <EuiTitle size="m">
+          <h2>{title}</h2>
+        </EuiTitle>
+      </EuiFlyoutHeader>
+      <EuiFlyoutBody>{children}</EuiFlyoutBody>
+      <EuiFlyoutFooter>
+        <EuiFlexGroup justifyContent="spaceBetween">
+          <EuiFlexItem grow={false}>
+            <EuiButtonEmpty flush="left" isDisabled={isLoading} onClick={onCancel}>
+              <FormattedMessage
+                id="xpack.security.formFlyout.cancelButton"
+                defaultMessage="Cancel"
+              />
+            </EuiButtonEmpty>
+          </EuiFlexItem>
+          <EuiFlexItem grow={false}>
+            <EuiButton
+              isLoading={isLoading}
+              isDisabled={isDisabled}
+              color={submitButtonColor}
+              fill
+              onClick={onSubmit}
+            >
+              {submitButtonText}
+            </EuiButton>
+          </EuiFlexItem>
+        </EuiFlexGroup>
+      </EuiFlyoutFooter>
+    </EuiFlyout>
+  );
+
+  return ownFocus ? <EuiPortal>{flyout}</EuiPortal> : flyout;
```

Comment:
So `ownFocus` seems to be a little overloaded here. I don't think wrapping in a portal should be solely restricted to the `ownFocus` prop (which adds the overlay mask). In fact, for best results when used in apps (Kibana) that use fixed positioned headers, they should really all be in a portal so they align best with those headers.

Reply:
Also, not sure if this is well known, but there is a global Kibana service for flyouts that maybe could be extended here. https://github.com/elastic/kibana/blob/master/src/core/public/overlays/flyout/flyout_service.tsx

Reply:
> In fact, for best results when used in apps (Kibana) that use fixed positioned headers, they should really all be in a portal so they align best with those headers.

I didn't know that was the recommendation - Let's add that to the EUI docs as a prominent code example. 

The only reason I added that condition was because `EuiFlyout` display is broken in Kibana when using the `ownFocus` prop unless rendered in a portal. (The overlay mask will get rendered on top of the flyout so you can't interact with it)

I think that might be the reason why most flyouts in Kibana are not rendered with an overlay mask right now, which makes it less of a focussed interaction. Definitely worth calling this out more explicitly. Some devs might not know how to use portals to fix those display issues.

Reply:
Yeah it's hard to make a prominent example in EUI because it's a unique situation of having fixed headers with flyouts. There's a lot of `if this, then that` situations that are hard to document "prominently". Which is why we'd like to lean into a global service that all plugin authors use to ensure the right configuration. _In fact, I think that service I referenced probably needs to force the portal as well._

There are cases where having the overlay mask is not wanted, like in the new Discover layout where the user can browse full documents while still viewing the full table below. Another reason why we try to make the UI components flexible in EUI but rely on Kibana to establish and reuse patterns specific to Kibana.